src/uct/ib/rc/accel/rc_mlx5_common.h

```diff
@@ -1297,4 +1301,107 @@ uct_rc_mlx5_iface_common_poll_rx(uct_rc_mlx5_iface_common_t *mlx5_common_iface,
     return count;
 }
 
+#if HAVE_IBV_EXP_DM
+/* DM memory should be written by 8 bytes (int64) to eliminate
+ * processor cache issues. To make this used uct_rc_mlx5_dm_copy_data_t
+ * datatype where first hdr_len bytes are filled by message header
+ * and tail is filled by head of message. */
+static void UCS_F_ALWAYS_INLINE
+uct_rc_mlx5_iface_common_copy_to_dm(uct_rc_mlx5_dm_copy_data_t *cache, size_t hdr_len,
+                                    const void *payload, size_t length, void *dm,
+                                    uct_ib_log_sge_t *log_sge)
+{
+    size_t head         = (cache && hdr_len) ? ucs_min(length, sizeof(*cache) - hdr_len) : 0;
+    size_t body         = ucs_align_down(length - head, sizeof(uint64_t));
+    size_t tail         = length - (head + body);
+    uint64_t *dst       = dm;
+    uint64_t padding    = 0; /* init by 0 to suppress valgrind error */
+    int i               = 0;
+
+    ucs_assert(sizeof(*cache) >= hdr_len);
+    ucs_assert(head + body + tail == length);
+    ucs_assert(tail < sizeof(uint64_t));
+
+    /* copy head of payload to tail of cache */
+    memcpy(cache->in + hdr_len, payload, head);
+
+    UCS_STATIC_ASSERT(sizeof(*cache) == sizeof(cache->out));
+    UCS_STATIC_ASSERT(sizeof(cache->in) == sizeof(cache->out));
+    UCS_STATIC_ASSERT(sizeof(log_sge->sg_list) / sizeof(log_sge->sg_list[0]) >= 2);
+
+    /* condition is static-evaluated */
+    if (cache && hdr_len) {
+        /* atomically by 8 bytes copy data to DM */
+        *(dst++) = cache->out[0];
+        *(dst++) = cache->out[1];
+        if (ucs_log_is_enabled(UCS_LOG_LEVEL_TRACE_DATA)) {
+            log_sge->sg_list[0].addr   = (uint64_t)cache;
+            log_sge->sg_list[0].length = (uint32_t)hdr_len;
+            i++;
+        }
+    }
+    if (ucs_log_is_enabled(UCS_LOG_LEVEL_TRACE_DATA)) {
+        log_sge->sg_list[i].addr   = (uint64_t)payload;
+        log_sge->sg_list[i].length = (uint32_t)length;
+        i++;
+    }
+    log_sge->num_sge = i;
+
+    /* copy payload to DM */
+    UCS_WORD_COPY(dst, payload + head, uint64_t, body);
+    if (tail) {
+        memcpy(&padding, payload + head + body, tail);
+        *(dst + (body / sizeof(uint64_t))) = padding;
+    }
+}
+
+static ucs_status_t UCS_F_ALWAYS_INLINE
+uct_rc_mlx5_common_make_data(uct_rc_mlx5_iface_common_t *iface,
```

Comment:
is it only for "dm" case? so the function name should reflect it

Reply:
renamed (added dm_ suffix)